keps/sig-api-machinery/2558-field-version-openapi/README.md

```diff
@@ -0,0 +1,414 @@
+and the feature gate associated with field.
+
+## Motivation
+
+**Feature gate**
+
+When a new field is added to an existing API version, a feature gate
+is added to control the enablement of the new field.
+
+The feature gates are defined in [`kube_features.go`]. The comments
+for each of these feature gates in [`kube_features.go`] lists
+the prerelease status of the feature gate and the minimum Kubernetes
+version for the particular prerelease. These comments do not follow
+a pattern and are not parseable.
+
+Currently, when a new field is added, there is no way to map and document
+the field with the feature gate that controls it. Some fields mention
+the feature gates in a comment but these comments are not parseable either.
+
+**Field version**
+
+Currently, when a new field is added to a stable API version, there
+is no way to expose the version of the new field in OpenAPI.
+One way to identify alpha API fields is by comparing the OpenAPI schema
+generated from an apiserver with all APIs enabled and the one generated
+from an apiserver with only stable and beta APIs enabled.
+However, this process is cumbersome and not always feasible because not all
+fields are gated at the API group level.
+
+Since the OpenAPI definition is the source of truth for API documentation,
+this KEP proposes a way to publish field-level versioning and feature gate
+information via OpenAPI.
+
+As a follow-up, this information can also be used to:
+- update the Kubernetes API reference documentation
+- surface [server-side warnings] in client-go and kubectl when a deprecated
+field or alpha field is being used.
+
+### Goals
+
+- Publish feature gate associated with an API field for built-in API types
+in OpenAPI to document the mapping of API fields with their feature gates.
+- Document the current prerelease status of an API field by publishing
+the prerelease status of the field and the minimum Kubernetes version for
+the prerelease in OpenAPI.
+- Validate that the feature gate mentioned in the API field comment
+is a valid feature gate.
+
+### Non-Goals
+
+- Support publishing field-level versioning for custom resources in OpenAPI.
+- Publish the default status of a feature gate (enabled/disabled by default)
+in OpenAPI.
+- Allow parsing comment tags on feature gates in [`kube_features.go`].
+- Update the Kubernetes API reference docs to include the field-level
+versioning information (out of scope of this KEP, can be done in a follow up).
+
+## Proposal
+
+This KEP proposes changes to the kube-openapi generator to add a new
+vendor extension field `x-kubernetes-field-version` in OpenAPI.
+This extension specifies the prerelease status of an API field,
+the minimum Kubernetes version for the prerelease and the
+feature gate associated with the field.
+
+This information is derived from comment tags on API fields with the format:
+
+```
+// +fieldLifecycle=<prerelease-status>
+// +fieldMinVersion=<k8s-version>
+// +fieldFeatureGate=<featuregate-name>
+```
```

Comment:
as we want to unify the marker language between in- and out of tree types, are we sure we want to introduce this kubernetism without qualifying them as being kubernetes related?

Reply:
What if I want to set a min version depending on OpenShift API or Vsphere?

Reply:
> What if I want to set a min version depending on OpenShift API or Vsphere?

Clarified that Kubernetes-specific constraints would be validated by a separate linter. PTAL.